### ISSUE TITLE
Cancelling the result of a merge does not cancel the merged observables

### ISSUE THREAD

**Issue by thSoft** [monixio/monix#142]
If we merge observables into a result observable, they should be canceled when the result is canceled. Am I right about whether this is the right semantic?

**Comment by alexandru** [comment]
@thSoft yes you're right. What version are you using?

**Comment by thSoft** [comment]
1.1 (AFAICS, the implementation is identical in 1.2 also)

**Comment by thSoft** [comment]
A similar issue in RxJava: https://github.com/ReactiveX/RxJava/issues/897

**Comment by alexandru** [comment]
Thanks for the bug, I will take a look.
 
Take note that in version 1.x the semantic of cancelation is different
from version 2.x or from RxJava. In 1.x streaming is strictly a
relationship between a producer and a consumer, which means that only
the consumer can cancel and it can do so only when receiving an onNext.
Hence cancel isn't seen immediately, but when the merged Observables
will emit their onNext, as that is when they can emit a Cancel.
Eventually they do get canceled though. So the Cancelable you get from
subscribe is basically built with .takeWhileNotCanceled()
 
The reason for this design choice has been that it makes things simpler.
By introducing a subscription that can be canceled at any point in time,
we introduce concurrency that is harder to manage. Plus in most cases it
can be worked around. But of course, there are always legitimate usages,
like in switchMap.
 
That is why version 2.0 will probably be what you want, as I changed the
notion of what cancel() is, aligning it with RxJava. It's at milestone 2
right now, but it's basically a release candidate. I want to declare it
final, but was waiting to see if bugs happen :-)
 
I am mobile right now and the above is just an assumption. I will double
check and also take a look at RxJava. If you could supply a simple code
sample reproducing your problem, that would be great.
 
--
Alexandru Nedelcu
bionicspirit.com
 
 
On Fri, Apr 22, 2016, at 09:46 AM, Dénes Harmath wrote:
> A similar issue in RxJava: ReactiveX/RxJava#897[1]
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub[2]
>
 

Links:

  1. https://github.com/ReactiveX/RxJava/issues/897
  2. https://github.com/monixio/monix/issues/142#issuecomment-213289610

**Comment by thSoft** [comment]
Thanks for the extensive info. Yes, I need this because I use switchMap and my merged observables are leaking. I'm going to supply a minimal example and try with 2.0-M2 as well.

**Comment by thSoft** [comment]
1.2:
```scala
import scala.concurrent.duration.DurationInt
import scala.scalajs.js.JSApp
import monifu.reactive.Observable
import monifu.concurrent.Implicits.globalScheduler

object Main extends JSApp {

  def main(): Unit = {
    Observable.interval(1 seconds).switchMap(i => {
      Observable.merge(
        Observable.interval(2 seconds).map(j => s"$i $j").doOnCanceled(println(s"$i canceled"))
      )
    }).take(4).foreach(println)
  }

}
```
Output (unexpected):
```
0 0
1 0
2 0
3 0
0 canceled
1 canceled
3 canceled
2 canceled
```

2.0-M2:
```scala
import scala.concurrent.duration.DurationInt
import scala.scalajs.js.JSApp
import monix.reactive.Observable
import monix.execution.Scheduler.Implicits._

object Main extends JSApp {

  def main(): Unit = {
    Observable.interval(1 seconds).switchMap(i => {
      Observable.merge(
        Observable.interval(2 seconds).map(j => s"$i $j").doOnCancel(println(s"$i canceled"))
      )
    }).take(4).foreach(println)
  }

}
```
Output (expected):
```
0 0
0 canceled
1 0
1 canceled
2 0
2 canceled
3 0
3 canceled
```

**Comment by alexandru** [comment]
@thSoft unfortunately for `1.x` that's expected behavior. Is the behavior from `2.x` satisfying?

Btw, I released `2.0-RC1` which is a feature freeze and contains a small bug fix for `refCount`.

**Comment by thSoft** [comment]
Oh, I understand. The 2.x behavior is totally satisfying. :) Thanks!
